addons/pos_self_order_razorpay/static/src/app/razorpay.js

```diff
@@ -0,0 +1,146 @@
+
+export class Razorpay {
+    constructor(...args) {
+        this.setup(...args);
+    }
+
+    setup(env, razorpayPaymentMethod, access_token, pos_config, errorCallback) {
+        this.env = env;
+        this.access_token = access_token;
+        this.razorpayPaymentMethod = razorpayPaymentMethod;
+        this.pos_config = pos_config;
+        this.errorCallback = errorCallback;
+        this.savedOrder = false;
+        this.pollTimeout = null;
+        this.inactivityTimeout = null;
+        this.queued = false;
+        this.payment_stopped = false;
+    }
+
+    handleRazorpayResponse(response) {
+        if (response?.error) {
+            this.payment_stopped
+                ? this.errorCallback(new RazorpayError("Transaction canceled due to inactivity"))
+                : this.errorCallback(new RazorpayError(response.error));
+            this.removePaymentHandler(["p2pRequestId"]);
+            return false;
+        }
+        localStorage.setItem("p2pRequestId", response?.p2pRequestId);
+        return true;
+    }
+
+    async cancelPayment(order) {
+        const data = { p2pRequestId: localStorage.getItem("p2pRequestId") };
+        try {
+            const cancel_response = await rpc("/pos-self-order/razorpay-cancel-transaction/", {
+                access_token: this.access_token,
+                order_id: order.id,
+                payment_data: data,
+                payment_method_id: this.razorpayPaymentMethod.id,
+            });
+            if (cancel_response) {
+                if (cancel_response?.errorMessage) {
+                    this.errorCallback(cancel_response.errorMessage, "warning");
+                    return true;
+                }
+                return this.handleRazorpayResponse(cancel_response);
+            }
+        } catch (error) {
+            this.errorCallback(error);
+            return false;
+        }
+    }
+
+    async startPayment(order) {
+        const call_razorpay = await this.processPayment(order);
+        if (call_razorpay) {
+            await this.paymentPolling(this.savedOrder);
+        }
+    }
+
+    async processPayment(order) {
+        try {
+            const initial_response = await rpc(`/kiosk/payment/${this.pos_config.id}/kiosk`, {
+                order: order.serialize({ orm: true }),
+                access_token: this.access_token,
+                payment_method_id: this.razorpayPaymentMethod.id,
+            });
+            if (initial_response) {
+                this.savedOrder = initial_response.order[0];
+                return this.handleRazorpayResponse(initial_response.payment_status);
+            }
+        } catch (error) {
+            this.errorCallback(error);
+            return false;
+        }
+    }
+
+    /**
+     * Polling
+     * This method calls and handles the razorpay payment status
+     * calls every 10 sec until payment status not found.
+     */
+    async paymentPolling(order) {
```

Comment:
Shouldn't you be using the websocket for this? Does razorpay notifiy the odoo server when a payment is done?

So when razorpay notifies the server, we send a websocket message that the payment is received. That message may contain the `p2pRequestId`. You will match that to the `p2pRequestId` (which I assumed to be) saved in the pos.payment object.

You may probably be able to avoid the use of the `localStorage` to persist the identity of the payment.

Reply:
@caburj, This was added because Razorpay lacks webhooks for POS, leaving with no option but to use long polling.